### PR TITLE
DCS-1033 New create booking flow

### PR DESCRIPTION
**Pull request by andrewrlee**
Previously there were a few steps to the wizard

date and time -> enter court and comment-> select rooms -> booking confirmed

Now we've gone down to:

enter booking details -> enter comment -> booking confirmed

This still requires changes to the availability checks - there's a commented out test that needs to be reinstated after we move to the new availability method

Also moved to Gurnank's new room endpoint!

New start page:
<kbd>
Screenshot 2021-06-30 at 21 28 33
</kbd>
Confirm page:
<kbd>
Screenshot 2021-06-30 at 21 28 42
</kbd>
Confirmation page:
<kbd>
Screenshot 2021-06-30 at 21 28 56
</kbd>

### CODE REVIEW

File: backend/api/prisonApi.ts [hunk 1]
```diff
@@ -59,12 +59,4 @@ export default class PrisonApi {
...
     return this.get(context, `/api/bookings/offenderNo/${offenderNo}?fullInfo=${fullInfo}`)
   }
-
-  public getLocation(context: Context, locationId: number): Promise<Location> {
-    return this.get(context, `/api/locations/${locationId}`)
-  }
-
-  public getLocationsForAppointments(context: Context, agencyId: string): Promise<Location[]> {
-    return this.get(context, `/api/agencies/${agencyId}/locations?eventType=APP`)
-  }
```
Comment (andrewrlee) [inline backend/api/prisonApi.ts:69]: no longer used

File: backend/routes/createBooking-v2/confirmBooking/commentValidation.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  commentLength: {
+    text: 'Maximum length should not exceed 3600 characters',
+    href: '#comment',
+  },
+}
+
+export default function validate(form: Record<string, unknown>): ValidationError[] {
+  assertHasOptionalStringValues(form, ['comment'])
+  const { comment } = form
+  return comment && comment.length > 3600 ? [errorTypes.commentLength] : []
```
Comment (andrewrlee) [inline backend/routes/createBooking-v2/confirmBooking/commentValidation.ts:14]: would really like to move the validation to [class validator](https://github.com/typestack/class-validator)

File: backend/services/bookingService.ts [hunk 3]
```diff
@@ -30,17 +29,13 @@ type AppointmentDetail = {
...
-  roomFinderFactory: RoomFinderFactory
-
   constructor(
     private readonly prisonApi: PrisonApi,
     private readonly whereaboutsApi: WhereaboutsApi,
     private readonly notificationService: NotificationService,
     private readonly availabilityCheckService: AvailabilityCheckService,
     private readonly locationService: LocationService
-  ) {
-    this.roomFinderFactory = roomFinderFactory(this.prisonApi)
```
Comment (andrewrlee) [inline backend/services/bookingService.ts:42]: moved this over to location service and it now uses whereabouts

File: backend/shared/dateAndTimeValidation.ts [hunk 4]
```diff
@@ -89,12 +89,14 @@ export default function validate(form: Record<string, unknown>): ValidationError
...
+    preRequired,
+    postRequired,
   } = form
 
   const errors: ValidationError[] = []
 
-  if (!preAppointmentRequired) errors.push(errorTypes.missingPreCourt)
-  if (!postAppointmentRequired) errors.push(errorTypes.missingPostCourt)
+  if (!preAppointmentRequired && !preRequired) errors.push(errorTypes.missingPreCourt)
+  if (!postAppointmentRequired && !postRequired) errors.push(errorTypes.missingPostCourt)
```
Comment (andrewrlee) [inline backend/shared/dateAndTimeValidation.ts:99]: this is a bit rubbish - this is shared across amend and create and the field is named different things in each flow. Will extract out in a follow up

File: integration-tests/integration/homePage.spec.js [hunk 5]
```diff
@@ -24,28 +23,15 @@ context('A user can view the video link home page', () => {
...
+      rooms: [],
     })
 
     cy.login()
   })
 
   it('should redirect a video court user to the video link home page', () => {
-    cy.task('stubLocationGroups')
-    cy.task('stubGroups', 'MDI')
-    cy.task('stubActivityLocations')
```
Comment (andrewrlee) [inline integration-tests/integration/homePage.spec.js:38]: not actually required

File: integration-tests/integration/homePage.spec.js [hunk 6]
```diff
@@ -6,12 +6,11 @@ const CourtVideoBookingsPage = require('../pages/viewBookings/courtVideoBookings
 const ManageCourtsPage = require('../pages/manageCourts/manageCourtsPage')
 
 context('A user can view the video link home page', () => {
-  before(() => {
```
Comment (andrewrlee) [inline integration-tests/integration/homePage.spec.js:9]: `before` breaks the UI and runs really slow

File: integration-tests/mockApis/prisonApi.js [hunk 7]
```diff
@@ -67,22 +67,6 @@ module.exports = {
     })
   },
 
-  stubAppointmentLocations: (agency, locations, status = 200) => {
```
Comment (andrewrlee) [inline integration-tests/mockApis/prisonApi.js:70]: no longer used

File: integration-tests/mockApis/whereabouts.js [hunk 8]
```diff
@@ -179,147 +197,6 @@ module.exports = {
...
-      },
-    ]
-
-    return stubFor({
-      request: {
-        method: 'GET',
-        url: `/whereabouts/agencies/${caseload.id}/locations/groups`,
-      },
-      response: {
-        status,
```
Comment (andrewrlee) [inline integration-tests/mockApis/whereabouts.js:284]: never used!

File: backend/routes/createBooking-v2/confirmBooking/commentValidation.test.ts [hunk 9]
```diff
@@ -0,0 +1,21 @@
+import validator, { errorTypes } from './commentValidation'
+
+describe('commentValdiation', () => {
```
Comment (bell-pa) [inline backend/routes/createBooking-v2/confirmBooking/commentValidation.test.ts:3]: Typo

File: backend/routes/createBooking-v2/confirmBooking/confirmBookingController.ts [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+        name: offenderNameWithNumber,
+        prison: agencyDetails.description,
+        court: court.text,
+      },
+      details: {
+        date: newBooking.date.format(DATE_ONLY_LONG_FORMAT_SPEC),
+        courtHearingStartTime: newBooking.startTime.format(MOMENT_TIME),
+        courtHearingEndTime: newBooking.endTime.format(MOMENT_TIME),
+        prisonRoomForCourtHearing: roomFinder.prisonRoom(newBooking.mainLocation),
+        'Pre-court hearing briefing': getPreDescription(newBooking.startTime, newBooking.preRequired),
```
Comment (bell-pa) [inline backend/routes/createBooking-v2/confirmBooking/confirmBookingController.ts:47]: This seems a bit convoluted.

Comment (andrewrlee) [reply backend/routes/createBooking-v2/confirmBooking/confirmBookingController.ts:47]: Might be due to the tight coupling of the data to the template - there's nunjucks macos that render this out raw as a summary list - so all the data needs to be properly formatted in the controller.

File: backend/routes/createBooking-v2/confirmBooking/commentValidation.test.ts (hunk 9 above)
Comment (andrewrlee) [reply backend/routes/createBooking-v2/confirmBooking/commentValidation.test.ts]: fixed

**Review by bell-pa — APPROVED** [review]
It _looks_ OK.
So far as I can tell.

**Pull request merged**